File: sdks/python/src/integrations/twilio.py
"""
Twilio Integration for the zv1 engine.

Provides SMS sending via the Twilio REST API.
"""

from __future__ import annotations

import time
from typing import Any

from src.utilities.sanitize_api_call import emit_api_call_event

try:
    import httpx
    HAS_HTTPX = True
except ImportError:
    HAS_HTTPX = False


class TwilioIntegration:
    def __init__(self, config: dict[str, str] | str, timeout: float = 30.0) -> None:
        if not HAS_HTTPX:
            raise ImportError("httpx package is required for Twilio integration.")
        if isinstance(config, dict):
            self.account_sid = config["account_sid"]
            self.auth_token = config["auth_token"]
        else:
            raise Exception("Twilio integration requires an object with account_sid and auth_token")
        self.base_url = f"https://api.twilio.com/2010-04-01/Accounts/{self.account_sid}"
        self.client = httpx.AsyncClient(
            timeout=timeout,
            auth=(self.account_sid, self.auth_token),
        )

    async def _request(self, method: str, url: str, data: dict[str, str] | None = None) -> dict[str, Any]:
        start_time = int(time.time() * 1000)
        try:
            response = await self.client.request(method, url, data=data)
            await emit_api_call_event(getattr(self, "_engine_config", None), {
                "timestamp": start_time, "integration": "twilio", "nodeId": None, "nodeType": None,
                "request": {"method": method, "url": url, "headers": {}, "body": data},
                "response": {"status": response.status_code, "statusText": response.reason_phrase or ""},
                "duration": int(time.time() * 1000) - start_time, "error": None,
            })
            if response.status_code >= 400:
                msg = response.json().get("message", response.reason_phrase) if response.content else response.reason_phrase
                raise Exception(f"Twilio API error ({response.status_code}): {msg}")
            return response.json()
        except Exception as e:
            if "Twilio API error" not in str(e):
                await emit_api_call_event(getattr(self, "_engine_config", None), {
                    "timestamp": start_time, "integration": "twilio", "nodeId": None, "nodeType": None,
                    "request": {"method": method, "url": url, "headers": {}, "body": data},
                    "response": {"status": 0, "statusText": "Error"},
                    "duration": int(time.time() * 1000) - start_time, "error": str(e),
                })
            raise

    async def send_sms(self, to: str, from_number: str, body: str) -> dict[str, Any]:
        return await self._request("POST", f"{self.base_url}/Messages.json", data={"To": to, "From": from_number, "Body": body})

    async def close(self) -> None:
        await self.client.aclose()

File: sdks/nodejs/src/integrations/twilio.js
import axios from 'axios';
import { emitAPICallEvent } from '../utilities/sanitizeAPICall.js';

export default class TwilioIntegration {
    constructor(config, options = {}) {
        if (typeof config === 'object') {
            this.accountSid = config.account_sid;
            this.authToken = config.auth_token;
        } else {
            throw new Error('Twilio integration requires an object with account_sid and auth_token');
        }
        this.options = {
            baseURL: `https://api.twilio.com/2010-04-01/Accounts/${this.accountSid}`,
            timeout: 30000,
            ...options
        };
    }

    async _request(config) {
        const requestHeaders = { 'Content-Type': 'application/x-www-form-urlencoded', ...config.headers };
        const startTime = Date.now();

        try {
            const response = await axios({
                ...config,
                headers: requestHeaders,
                auth: { username: this.accountSid, password: this.authToken },
                timeout: this.options.timeout
            });

            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'twilio', nodeId: null, nodeType: null,
                request: { method: config.method, url: config.url, headers: requestHeaders, body: config.data || null },
                response: { status: response.status, statusText: response.statusText },
                duration: Date.now() - startTime, error: null
            });

            return response.data;
        } catch (error) {
            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'twilio', nodeId: null, nodeType: null,
                request: { method: config.method, url: config.url, headers: requestHeaders, body: config.data || null },
                response: { status: error.response?.status || 0, statusText: error.response?.statusText || 'Error' },
                duration: Date.now() - startTime, error: error.message
            });
            if (error.response) {
                const msg = error.response.data?.message || error.response.statusText;
                throw new Error(`Twilio API error (${error.response.status}): ${msg}`);
            }
            throw new Error(`Twilio API error: ${error.message}`);
        }
    }

    async sendSMS(to, from, body) {
        const params = new URLSearchParams();
        params.append('To', to);
        params.append('From', from);
        params.append('Body', body);

        return await this._request({
            method: 'POST',
            url: `${this.options.baseURL}/Messages.json`,
            data: params.toString()
        });
    }
}
